modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java

```diff
@@ -539,31 +543,31 @@ public boolean equals(Object obj) {
     private boolean equalsBitSet(BitSet<?> other) {
         int a = this.bits != null ? this.bits.length : 0;
         int b = other.bits != null ? other.bits.length : 0;
+        int max = Math.max(a, b);
 
-        if (a != b) return false;
```

Comment:
I understand the following loop will return false in case a != b as well, but for performance reasons, wouldn't it be best to keep this test (as it immediately fails in case a != b)?

Reply:
The problem is that `BitSet` isn't diligent enough about how it manages its arrays.  If all bits get cleared in the highest `long`, it would need to shrink the array (potentially by more than one `long` if the next highest `long` is also `0`), but of course it doesn't.  The only thing it does is to reset to an empty array when **all** bits were cleared.

So `a != b` does not guarantee they're not equal.  If `a` has length 1 and `b` has length 2 but its `bits[1]` is `0`, they're still equal (assuming `bits[0]` is equal for both).

The whole class is poorly engineered unfortunately -- note that almost all comments about this change are pointing out problems in `BitSet`, not about the actual caching.

I think it would best to remove `BitSet`.  The need for it is minimal once the caching goes in; saving space when you have a million copies of these is useful, but no longer relevant when they're now a million references to the same (immutable) copy.

Reply:
> I think it would best to remove `BitSet`

does it mean there'll be a followup PR?  or should it be done as a part of this one?

Reply:
> > I think it would best to remove BitSet
> 
> does it mean there'll be a followup PR?

Filing a follow-up cleanup bug seems reasonable, but not urgent, especially now that BitSet is immutable . The problems pointed out are (largely, if not wholly) theoretical, but it does continue to be a maintenance burden.

> or should it be done as a part of this one?

Definitely not.